### PR TITLE
[mlir][LLVM] Add operand bundle support

### PR DESCRIPTION
**Pull request by Lancern**
This PR adds LLVM [operand bundle](https://llvm.org/docs/LangRef.html#operand-bundles) support to MLIR LLVM dialect. It affects these 3 operations related to making function calls: `llvm.call`, `llvm.invoke`, and `llvm.call_intrinsic`.

This PR adds two new parameters to each of the 3 operations. The first parameter is a variadic operand `op_bundle_operands` that contains the SSA values for operand bundles. The second parameter is a property `op_bundle_tags` which holds an array of strings that represent the tags of each operand bundle.

### CODE REVIEW

**Review by zero9178 — COMMENTED** [review]
Is there a reason we cannot use [`VariadicOfVariadic`](https://github.com/llvm/llvm-project/blob/019136e30fea8b8e6ef9c6dda1f9dd20a504c573/mlir/include/mlir/IR/CommonTypeConstraints.td#L135) for `bundle_operands`?
`op_bundles` would then need to be a `DenseI32ArrayAttr`, but ought to be kept up to sync automatically by all APIs I think.

**Comment by Lancern** [comment]
> Is there a reason we cannot use [VariadicOfVariadic](https://github.com/llvm/llvm-project/blob/019136e30fea8b8e6ef9c6dda1f9dd20a504c573/mlir/include/mlir/IR/CommonTypeConstraints.td#L135) for bundle_operands?
op_bundles would then need to be a DenseI32ArrayAttr, but ought to be kept up to sync automatically by all APIs I think.

Each operand bundle has a string tag associated with it, thus I have to invent the `#llvm.opbundle` stuff to store this tag. Do you suggest that the string tags should be kept separately?

**Comment by zero9178** [comment]
> Do you suggest that the string tags should be kept separately?

I believe this should result in less code and things that could go out of sync. Keeping the number of operands in the current `Variadic` in sync with the indices used in the operand bundles rather difficult and the implementation detail of the ranges being denoted via indices even leaks into the assembly format. 

With the `VariadicOfVariadic` the only thing that needs to be kept in sync is the number of variadic ranges and the number of corresponding string tags. We could then write a custom printer and parser that prints each operand range together with its string tag.

**Comment by Lancern** [comment]
@zero9178 I'll update to follow your proposed approach.

**Comment by Lancern** [comment]
@zero9178 I've updated to use `VariadicOfVariadic` for `op_bundle_operands`.

File: mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp
Comment (zero9178) [inline mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp:549]: Why is this necessary? The builder call should create this automatically from its arguments IIRC (and if it doesn't, then `build` call, if implemented manually, should be updated)

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (zero9178) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:560]:
```suggestion
                     ArrayProperty<StringProperty, "operand bundle tags">:$op_bundle_tags);
```
Not sure I understand why we need to wrap this in an optional. Is there a semantic difference between an empty array and the array being non-present?

File: mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp
Comment (zero9178) [inline mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp:551]: This should not necessary if the `OptionalProperty` is deleted.
(and would otherwise better be spelled as `newOp.properties.op_bundles_sizes = {};` I believe)

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (zero9178) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:1916]: Not sure how others feel about this, but I think I'd be nice to align the syntax relatively close to LLVM IR syntax and less error prone. Changing the syntax later would be somewhat difficult as well, so we probably want to get this right from the beginning.

What are your thoughts about the syntax:
```
operand bundle set ::= '[' operand bundle (, operand bundle )* ']'
operand bundle ::= $op_bundle_tags[i] '(' $op_bundle_operands[i] ':' type($op_bundle_operands[i]) ')'
```
I would place this right after the call arguments.
We could achieve this here in the assembly format using:
```suggestion
    $intrin `(` $args `)`
    (custom<OpBundles>($op_bundle_operands, $op_bundle_tags)^)?
    `:` functional-type($args, $results)
```
The `printOpBundles` and `parseOpBundles` (returning `OptionalResult`) functions that will need to be implemented can also be used in the implementations of call and invoke op for their syntax.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1369]: Can this be rewritten using `parseCommaSeparatedList`? It has an overload taking delimiters that handles these common cases.

**Review by zero9178 — COMMENTED** [review]
Thank you so much! This is pretty much as I imagined conceptually.

I still have one more high-level design question about the syntax before I will do a more thorough pass over the actual code, though I did leave some comments in the code where I noticed issues.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:1916]: I am supportive of @zero9178 's proposal. It seems to simplify things and sticking to LLVM syntax is a plus.

File: mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp
Comment (Lancern) [reply mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp:549]: The `build` call is calling into an auto-generated overload that builds an operation all from scratch:

```cpp
static void build(
  ::mlir::OpBuilder &,
  ::mlir::OperationState &odsState,
  ::mlir::TypeRange resultTypes,
  ::mlir::ValueRange operands,
  ::llvm::ArrayRef<::mlir::NamedAttribute> attributes = {}
);
```

This overload does not perform any "automatic inference" of operands or attributes or something and we have to set the attribute manually.

The same for the change in `SPIRVToLLVM.cpp`.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (Lancern) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:560]: Because `LLVM_CallIntrinsicOp` uses declarative assembly format and `OptionalProperty` could prevent `op_bundle_tags` from appearing in the assembly if it's non-present. Anyway as we're going to change the assembly syntax, I'll create a custom printer and parser for `LLVM_CallIntrinsicOp` and the `OptionalProperty` could be removed then.

File: mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp
Comment (Lancern) [reply mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp]: Updated.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (Lancern) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td]: Updated to use this syntax.

**Comment by Lancern** [comment]
@zero9178 Hi I've updated the PR to use the proposed syntax.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1427]:
```suggestion
  for (const auto &operands : opBundleOperands) 
    opBundleSizes.push_back(operands.size());
```
nit:

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1413]:
```suggestion
       llvm::zip_equal(opBundleOperands, opBundleOperandTypes)) {
```
nit: zip_equal should assert if the lengths don't match so you may use it to replace the assert above.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1442]: nit: this seems outdated now?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:256]:
```suggestion
  SMLoc currentParserLoc = p.getCurrentLocation();
```
expand the auto here

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1200]:
```suggestion
  ArrayRef<std::string> opBundleTags = op.getOpBundleTags();

  if (opBundleTags.size() != opBundleOperands.size())
```

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1490]:
```suggestion
  SMLoc opBundlesLoc = parser.getCurrentLocation();
```

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1491]: Expand the auto here

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1493]:
```suggestion
      result && failed(*result))
```
I think in LLVM/MLIR its more conventional to rely on implicit/contextual bool conversion

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp
Comment (zero9178) [inline mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:110]: Expand the auto here

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp
Comment (zero9178) [inline mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:134]:
```suggestion
  if (!bundleTags)
```

File: mlir/test/Dialect/LLVMIR/invalid.mlir
Comment (zero9178) [inline mlir/test/Dialect/LLVMIR/invalid.mlir:1665]:
```suggestion
  llvm.call @foo() ["tag"(%0 : i32, i32)] : () -> ()
```

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (zero9178) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:275]: I suspect that we need logic here to handle empty operand lists.
Specifically you'll want to do a `parseOptionalRParen` right after parsing the left paren to then skip the operand list entirely.

Could you also add a test for this case?

**Review by zero9178 — COMMENTED** [review]
Perfect, thank you so much! I added some nits regarding code-style and found oen case that I believe is a bug. Otherwise I think this will soon be good to go

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (gysit) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2665]: There is also an "llvm.intr.assume" intrinsic that would benefit from operand bundle support. It is defined in LLVMIntrinsicOps.td. I wonder if this could be updated as well? AFAIK most intrinsics are nowadays represented using specific operations rather than with the generic llvm.call_intrinsic.

**Review by gysit — COMMENTED** [review]
Thanks for working on this!

I added some mostly minor comments. It seems like some verifier tests and probably also some roundtrip.mlir tests for printing and parsing could make sense?

If you have time to add support for the llvm.intr.assume that would be great.

I think mid term we also want to support operand bundles when importing from LLVM IR into MLIR LLVM dialect (ModuleImport.cpp). However, given the size of the PR it may make sense to do that in a later stage.

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (Lancern) [reply mlir/test/Target/LLVMIR/llvmir.mlir:2665]: I'll make another PR for `llvm.intr.assume` after this PR lands. After all this PR is about operand bundles :)

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:275]: Test added.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
Comment (Lancern) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:1193]: Added a test for `llvm.call`.

**Comment by Lancern** [comment]
@zero9178 @gysit I've updated the code again to resolve the nits and add more tests.

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (zero9178) [inline mlir/test/Target/LLVMIR/llvmir.mlir:2637]: Could you also test:
```mlir
  llvm.call @foo() ["tag"()] : () -> ()
```

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:125]: nit: We may use zip_equal here as well except that assert message will help debugging in the future.

**Review by gysit — COMMENTED** [review]
I think it is worth having tests in roundtrip.mlir since the printing and parsing is non-trivial here?

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (Lancern) [reply mlir/test/Target/LLVMIR/llvmir.mlir:2637]: Updated the parser and added the test.

**Review by zero9178 — APPROVED** [review]
LGTM once @gysit's comments are addressed and he is happy :slightly_smiling_face:

File: mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp
Comment (Lancern) [reply mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp]: Updated.

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (Lancern) [reply mlir/test/Target/LLVMIR/llvmir.mlir:2637]: > Given the complexity of the printing and parsing it may also make sense to have some tests in the roundtrip.mlir to test the happy case?

This indeed caught some bugs in the printer for `llvm.invoke`! Updated.

**Review by gysit — APPROVED** [review]
Thanks for the tests! LGTM

File: mlir/test/Target/LLVMIR/llvmir.mlir
Comment (gysit) [reply mlir/test/Target/LLVMIR/llvmir.mlir:2637]: Yeah from experience printing and parsing is quite error prone...

**Comment by Lancern** [comment]
Rebased onto the latest `main`.

**Comment by Lancern** [comment]
BTW I don't have commit access so if all reviewers are happy about this PR someone needs to land it.

**Comment by Dinistro** [comment]
There seems to be a flang issue caused by producing an invalid `llvm.call` operation.

**Comment by gysit** [comment]
Once CI is green one of us will happily click the button.

As this is not your first contribution - and since we hope for many more :) - you can consider to apply for commit access (https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access) once this landed.

**Comment by Lancern** [comment]
> There seems to be a flang issue caused by producing an invalid llvm.call operation.

Fixed these regressions in my local environment. Let's see if the CI passes.

> As this is not your first contribution - and since we hope for many more :) - you can consider to apply for commit access (https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access) once this landed.

Great, I'll apply for the commit access after this PR lands!

**Comment by Lancern** [comment]
The CI is green now.

**Comment by antiagainst** [comment]
Are the operand bundle a very core part of the llvm call ops? IIUC it's just auxiliary data? I suppose most of the time a user won't need to specify anything there. For such cases it would be nice to have some default builder or helper functions to make it transparent for downstream users. Right now it's kinda forcing all downstream users to add their wrappers to just ignore these newly added fields, e.g., https://github.com/triton-lang/triton/pull/4847.

**Comment by Lancern** [comment]
@antiagainst Hi Zhang, thanks for your feedback!

`llvm.call` and `llvm.invoke` already have some convenient builders. For `llvm.call`:
https://github.com/llvm/llvm-project/blob/1e5e153485b817422c311f9326e80781a83ea7bc/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td#L685-L700

For `llvm.invoke`:
https://github.com/llvm/llvm-project/blob/1e5e153485b817422c311f9326e80781a83ea7bc/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td#L567-L576

All of the above custom builders set the relevant attributes to their default values and you don't have to specify the operand bundle related attributes by hand if you use these builders.

I noticed that your code uses this builder overload:

```cpp
static void build(mlir::OpBuilder &, mlir::OperationState &odsState, mlir::TypeRange resultTypes, mlir::ValueRange operands, llvm::ArrayRef<mlir::NamedAttribute> attributes = {});
```

Implementation of this builder overload is generated automatically by MLIR and it lets you build the operation with every detail by hand. If you're using this overload you have to specify the operand bundle related attributes manually. I'm afraid we couldn't do much about this since this builder is generated by MLIR.

However `llvm.call_intrinsics` indeed does not have a custom builder that avoids the burden of specifying operand bundle attributes. We definitely could improve this!

**Comment by nicolasvasilache** [comment]
@Lancern why are these additions not marked optional in the op definition?

**Pull request merged**
